certora/specs/AssetsAccounting.spec

```diff
@@ -0,0 +1,126 @@
+    uint256 totalBorrowShares = virtualTotalBorrowShares(id);
+
+    return libMulDivUp(userShares, totalBorrowAssets, totalBorrowShares);
+}
+
+// Check that the assets supplied are greater than the assets owned in the end.
+rule supplyAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total supply assets.
+    require lastUpdate(id) == e.block.timestamp;
+    // Assume no supply position to begin with.
+    require supplyShares(id, onBehalf) == 0;
+
+    uint256 suppliedAssets;
+    suppliedAssets, _ = supply(e, marketParams, assets, shares, onBehalf, data);
+
+    uint256 ownedAssets = ownedSupplyAssets(id, onBehalf);
+
+    assert suppliedAssets >= ownedAssets;
+}
+
+// Check that the assets withdrawn are less than the assets owned initially.
+rule withdrawAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total supply assets.
+    require lastUpdate(id) == e.block.timestamp;
+
+    uint256 ownedAssets = ownedSupplyAssets(id, onBehalf);
+
+    uint256 withdrawnAssets;
+    withdrawnAssets, _ = withdraw(e, marketParams, assets, shares, onBehalf, receiver);
+
+    assert withdrawnAssets <= ownedAssets;
+}
+
+// Check that the assets borrowed are less than the assets owed in the end.
+rule borrowAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 shares, address onBehalf, address receiver) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total borrowed assets.
+    require lastUpdate(id) == e.block.timestamp;
+    // Assume no outstanding debt to begin with.
+    require borrowShares(id, onBehalf) == 0;
+
+    // The borrow call is restricted to shares as input to make it easier on the prover.
+    uint256 borrowedAssets;
+    borrowedAssets, _ = borrow(e, marketParams, 0, shares, onBehalf, receiver);
+
+    uint256 owedAssets = owedBorrowAssets(id, onBehalf);
+
+    assert borrowedAssets <= owedAssets;
+}
+
+// Check that the assets repaid are greater than the assets owed initially.
+rule repayAssetsAccounting(env e, MorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) {
+    MorphoHarness.Id id = libId(marketParams);
+
+    // Assume no interest as it would increase the total borrowed assets.
+    require lastUpdate(id) == e.block.timestamp;
+
+    uint256 owedAssets = owedBorrowAssets(id, onBehalf);
+
+    uint256 repaidAssets;
+    repaidAssets, _ = repay(e, marketParams, assets, shares, onBehalf, data);
+
+    // Assume a full repay.
+    require borrowShares(id, onBehalf) == 0;
+
+    assert repaidAssets >= owedAssets;
+}
+
```

Comment:
we might be able to do the same rule for liquidatoin

Reply:
Not sure this is the same thing: in this file we are comparing different ways to value a position. For example, for the supply side, we compare: 
- how much it did cost you in terms of loan assets to enter a supply position
- how much you can retrieve in terms of loan assets when exiting a supply position
- an arbitrarily defined function to estimate supply assets

For the liquidations though, I don't think it defines what your borrow/collateral position is. This is just something that can happen to your position, and we could make it worse or better without impacting what you own. For example, if we had a close factor of 50%, it would not mean that your collateral is worth twice as less